### PR TITLE
Stop on error

### PR DESCRIPTION
**Pull request by winduptoy**
Added `-e` flag to `rset`, which stops execution if an error is encountered within a label. New warning printed if an error is encountered within a label.

When running labels with `/bin/sh -x`, there is typically *a lot* of output, and with many labels it is unlikely that I would necessarily see errors or know where to begin looking for one. I typically add the `-e` flag to `/bin/sh` too, which is the primary case here. The red warning printed to the console helps with that, and stopping immediately leaves the error right in front of me on the terminal rather than way up in the scrollback. My personal preference is that this would be enabled by default. Thoughts?

### CODE REVIEW

File: rset.c [hunk 1]
```diff
@@ -301,10 +306,26 @@ int main(int argc, char *argv[])
 					hl_range(host_labels[j]->name, HL_LABEL, 0, 0);
 					printf("\n");
 				}
-				if (tty_opt)
-					(void) ssh_command_tty(hostname, socket_path, host_labels[j], http_port);
-				else
-					(void) ssh_command_pipe(hostname, socket_path, host_labels[j], http_port);
+
+				int exit_code;
```
Comment (eradman) [inline rset.c:310]: Move this to the top where the other `int`'s are defined (sorted alphanumerically)

File: rset.c [hunk 2]
```diff
@@ -301,10 +306,26 @@ int main(int argc, char *argv[])
...
+				if (tty_opt) {
+					exit_code = ssh_command_tty(hostname, socket_path, host_labels[j], http_port);
+				} else {
+					exit_code = ssh_command_pipe(hostname, socket_path, host_labels[j], http_port);
```
Comment (eradman) [inline rset.c:314]: Let's leave off the braces for these single statements. (I may change my mind about this later, just a preference)

File: rset.c [hunk 3]
```diff
@@ -301,10 +306,26 @@ int main(int argc, char *argv[])
...
+				}
+				if (exit_code != 0) {
+					snprintf(buf, sizeof(buf), "Exited with code %d", exit_code);
```
Comment (eradman) [inline rset.c:317]: Let's show the interpreter invocation that returned this error

```c
Options toplevel_options, op;
/*

*/
apply_default(op.interpreter, host_labels[j]->options.interpreter, INTERPRETER);
snprintf(buf, sizeof(buf), "%s exited with code %d", op.interpreter, exit_code);
```

File: rset.c [hunk 4]
```diff
@@ -301,10 +306,26 @@ int main(int argc, char *argv[])
...
+					hl_range(buf, HL_ERROR, 0, 0);
+					printf("\n");
+				}
+				if ((stop_on_err_opt) && exit_code != 0) {
```
Comment (eradman) [inline rset.c:321]: Let's use a `goto`

```c
if ((stop_on_err_opt) && exit_code != 0)
    goto exit;
```
...

```c
    if (stop_on_err_opt)
        return exit_code;
    else
        return 0;
```

File: rset.c [hunk 5]
```diff
@@ -90,7 +91,7 @@ int main(int argc, char *argv[])
 	char *sshconfig_file = NULL;
 
 	opterr = 0;
-	while ((ch = getopt(argc, argv, "lntvF:f:x:")) != -1)
+	while ((ch = getopt(argc, argv, "lntevF:f:x:")) != -1)
```
Comment (eradman) [inline rset.c:94]: List options alphabetically

**Review by eradman — CHANGES_REQUESTED** [review]
This is a useful change.  I'm not sure if the error should be printed if `-e` is not set--
Borne shells relay the exit code of the last command, which will mean a lot of false warnings for some.

File: rset.c (hunk 4 above)
Comment (winduptoy) [reply rset.c]: Can you verify my logic here in the new commit? I think `exit` needs to be above the `socket_path` cleanup code.

**Comment by winduptoy** [comment]
@eradman thanks for the review!

**Comment by eradman** [comment]
Thanks for the changes, merged.  There isn't a good place to write an automated test for this code.  I'll think about how to do that.

**Pull request merged**